### PR TITLE
Feat/cg delete case/cases

### PR DESCRIPTION
**Pull request by patrikgrenfeldt**
This PR adds functionality to delete cases via the CLI, e.g:
- cg delete case funnyduck
- cg delete cases --sample-identifier ticket_number 123456

### How to prepare for test
- [x] ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] paxa the environment: `paxa`
- [x] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh [THIS-BRANCH-NAME]`

### How to test deletion of case
- [x] do cg delete case [CASE-ID] for a case without analyses or has being processed in any way
- [x] answer yes to all questions

### Expected test outcome
- [x] check that the case was deleted from the database

### How to test dry-run deletion of case
- [x] do cg delete case [CASE-ID] --dry-run for a case without analyses or has being processed in any way
- [x] answer yes to all questions

### Expected test outcome
- [x] check that the case was NOT deleted from the database

### How to test 'no confirmation' when deleting a case
- [x] do cg delete case [CASE-ID] --yes for a case without analyses or has being processed in any way

### Expected test outcome
- [x] check that there was no question to answer yes to and that the case was deleted from the database

### How to test deleting multiple cases
- [x] do cg delete cases --sample-identifier ticket_number 123456 

### Expected test outcome
- [x] check that the delete case functionality was invoked for at least the first case
- [x] Take a screenshot and attach or copy/paste the output.

## Review
- [x] code approved by MR
- [x] tests executed by MR, PG
- [x] "Merge and deploy" approved by PG
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan
- [x] Document in code
- [ ] Deploy this branch
- [ ] Inform to lab and production

### CODE REVIEW

File: tests/cli/delete/test_cases.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    # WHEN calling delete cases with valid sample identifiers
+    result = cli_runner.invoke(
+        cases,
+        ["--sample-identifier", identifier_key, identifier_value],
+        obj=base_context,
+    )
+
+    # THEN it should name the case to be changed
+    assert case.internal_id in caplog.text
+    assert case.name in caplog.text
```
Comment (Mropat) [inline tests/cli/delete/test_cases.py:35]: Also check that there is no case in store in this test!

File: tests/cli/delete/test_case.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+def test_delete_case_bad_case(cli_runner, base_context):
+    """Test to delete a case using a non-existing case """
+    # GIVEN an empty database
+
+    # WHEN deleting a case
+    case_id = "dummy_name"
+    result = cli_runner.invoke(case, [case_id], obj=base_context)
+
+    # THEN then it should complain on missing case
+    print(result.output)
```
Comment (Mropat) [inline tests/cli/delete/test_case.py:31]: dont forget this :stuck_out_tongue:

File: tests/cli/delete/test_case.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    """Test that the delete case can't delete a case with analysis"""
+    # GIVEN a database with a case with an analysis
+    analysis_obj = helpers.add_analysis(base_store)
+    case_id = analysis_obj.family.internal_id
+
+    # WHEN deleting a case
+
+    result = cli_runner.invoke(case, [case_id, "--yes"], obj=base_context)
+
+    # THEN then it should have been deleted
```
Comment (Mropat) [inline tests/cli/delete/test_case.py:61]:
```suggestion
    # THEN then it should not have been deleted
```

File: tests/cli/delete/test_case.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    case_id = case_obj.internal_id
+    assert not case_obj.links
+
+    # WHEN deleting a case
+
+    result = cli_runner.invoke(case, [case_id, "--yes", "--dry-run"], obj=base_context)
+
+    # THEN then it should not have been deleted
+    assert result.exit_code == SUCCESS
+    assert base_store.Family.query.count() == 1
```
Comment (Mropat) [inline tests/cli/delete/test_case.py:79]: Perhaps check for dry-run specific output?

**Comment by Mropat** [comment]
How are you going to handle deleting of cases with samples which are present in other cases? Perhaps you have to check for each sample in case, that number of links is 1, otherwise not delete sample?

**Comment by patrikgrenfeldt** [comment]
> How are you going to handle deleting of cases with samples which are present in other cases? Perhaps you have to check for each sample in case, that number of links is 1, otherwise not delete sample?

I think the check is there already, in the backref sample.links

**Comment by annaengstrom** [comment]
@patrikgrenfeldt This doesn't seems to involve delete sample, is that another PR?

File: tests/cli/delete/test_case.py (hunk 2 above)
Comment (patrikgrenfeldt) [reply tests/cli/delete/test_case.py:31]: removed

File: tests/cli/delete/test_case.py (hunk 4 above)
Comment (patrikgrenfeldt) [reply tests/cli/delete/test_case.py:79]: I think there is no dry-run specific output.. should I implement some

**Comment by patrikgrenfeldt** [comment]
> @patrikgrenfeldt This doesn't seems to involve delete sample, is that another PR?

If there are samples in the case/family they will be deleted as well as long as they aren't used in other cases/families.
There is yes no specific functionality to remove sample(s) by themselves. As we have implemented cg now each sample is usually connected to a case/family. 

I.e. 'cg delete cases --sample-identifier ticket_number 123456' will delete each FamilySample, each Family as well as each Sample with the given ticket-number

With the above in mind, how pressing is it to get the "delete sample" functionality in place?

**Comment by patrikgrenfeldt** [comment]
Test result delete case:
image
🆗 

Test result dry-run: 
image
🆗 

Test --yes:
image
🆗 

Test delete cases:
image
🆗 

🏁

File: tests/cli/delete/test_case.py (hunk 4 above)
Comment (Mropat) [reply tests/cli/delete/test_case.py:77]: Yes! This way it will be clear if its dry-run or real command!

**Review by Mropat — APPROVED** [review]
:+1: 

Please fix the code smells and tests before deploying! Otherwise great work! :rocket: :rocket:

File: tests/cli/delete/test_case.py (hunk 4 above)
Comment (patrikgrenfeldt) [reply tests/cli/delete/test_case.py:83]: implemented now

**Comment by patrikgrenfeldt** [comment]
Deployed to Hasta:

image

**Comment by annaengstrom** [comment]
We would also like to be able to delete samples in a case where other samples remain, thus case should remain.

**Comment by patrikgrenfeldt** [comment]
If it is just a sample or two I guess it could easily be done from within clinical-api? But yes, I would like to implement this :) it's just a little bit down on the TODO list at the moment

**Comment by annaengstrom** [comment]
> If it is just a sample or two I guess it could easily be done from within clinical-api? But yes, I would like to implement this :) it's just a little bit down on the TODO list at the moment

@patrikgrenfeldt We can do it manually as well, but would nice to have it implemented later when time allows

**Pull request merged**